File: kotlin/src/test/kotlin/com/peshchuk/topjslibs/TopJsLibrariesTest.kt
package com.peshchuk.topjslibs

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.newSingleThreadContext
import kotlinx.coroutines.runBlocking
import kotlinx.coroutines.test.resetMain
import kotlinx.coroutines.test.setMain
import kotlin.test.AfterTest
import kotlin.test.BeforeTest
import kotlin.test.Test
import kotlin.test.assertEquals

@Suppress("EXPERIMENTAL_API_USAGE")
internal class TopJsLibrariesTest {

    private val mainThreadSurrogate = newSingleThreadContext("Test thread")

    @BeforeTest
    fun setUp() {
        Dispatchers.setMain(mainThreadSurrogate)
    }

    @AfterTest
    fun tearDown() {
        Dispatchers.resetMain() // reset main dispatcher to the original Main dispatcher
        mainThreadSurrogate.close()
    }

    @Test
    fun count() {
        runBlocking {
            launch(Dispatchers.Main) {
                // Arrange
                fun fetchHtml(url: String): String {
                    return when (url) {
                        "https://www.google.com/search?q=test" -> Util.getHtmlFrom("google.com.html")

                        "https://www.test.de/" -> Util.getHtmlFrom("test.de.html")
                        "https://www.test.de/shop/test-hefte/" -> Util.getHtmlFrom("test-hefte.html")
                        "https://www.test.de/thema/" -> Util.getHtmlFrom("thema.html")
                        "https://www.oekotest.de/" -> Util.getHtmlFrom("oekotest.de.html")
                        "https://de.wikipedia.org/wiki/Test_(Zeitschrift)" -> Util.getHtmlFrom("zeitschrift.html")

                        else -> ""
                    }
                }

                val expected = linkedMapOf(
                    Pair("base.min;v35893530.js", 2),
                    Pair("modules.min;v35893530.js", 2),
                    Pair("webtk.min;v35893530.js", 2),
                    Pair("webtrekk_cookieControl.min;v35893530.js", 2),
                    Pair("application-0d7a7bcfe504533ad327.js", 1)
                )
                val topJsLibraries =
                    TopJsLibraries(linksLimit = 5, topLimit = 5) { url -> fetchHtml(url) }

                // Act
                val actual = topJsLibraries.count("test")

                // Assert
                assertEquals(expected, actual)
            }
        }
    }
}

File: go/topjslibs/top_js_libraries_test.go
package topjslibs

import (
	"testing"
)

func TestCount(t *testing.T) {
	// Arrange
	expected := [5]JsNum{
		{"base.min;v35893530.js", 2},
		{"modules.min;v35893530.js", 2},
		{"webtk.min;v35893530.js", 2},
		{"webtrekk_cookieControl.min;v35893530.js", 2},
		{"application-0d7a7bcfe504533ad327.js", 1},
	}

	// Act
	topJsLibraries := TopJsLibraries{LinksLimit: 5, TopLimit: 5, FetchHtml: fetchHtml}
	var actual [5]JsNum
	copy(actual[:], topJsLibraries.Count("test"))

	// Assert
	if expected != actual {
		t.Errorf("TopJsLibraries{LinksLimit: 5, TopLimit: 5}.count(\"test\") = %s; want %s", actual, expected)
	}
}

func fetchHtml(url string) string {
	switch url {
	case "https://www.google.com/search?q=test":
		return getHtmlFrom("google.com.html")

	case "https://www.test.de/":
		return getHtmlFrom("test.de.html")
	case "https://www.test.de/shop/test-hefte/":
		return getHtmlFrom("test-hefte.html")
	case "https://www.test.de/thema/":
		return getHtmlFrom("thema.html")
	case "https://www.oekotest.de/":
		return getHtmlFrom("oekotest.de.html")
	case "https://de.wikipedia.org/wiki/Test_(Zeitschrift)":
		return getHtmlFrom("zeitschrift.html")

	default:
		return ""
	}
}
